### DISCUSSION TITLE
Duda de implementacion de enums con comportamiento

### DISCUSSION THREAD

**Discussion by Franpanozzo** [dds-jv/foro#229] [Consultas de objetos] [answered]
Buenas, al momento de revisar la iteracion 2 se me ocurrio cambiar algunas cosas y me estoy encontrando con un error y no sabria que estoy haciendo mal. 
`public enum TipoPrenda {
  REMERA {
    public Enum<Categoria> getCategoria() {
      return Categoria.PARTESUP;
    }
`
este enum tiene el metodo abstracto: `public abstract Enum<Categoria> getCategoria();`

pero al momento de querer llamarlo desde otra clase: 
`public Enum<Categoria> getCategoria(){
    return tipoPrenda.getCategoria();
  }`
Tira error _Cannot resolve method 'getCategoria' in 'Enum'_

tampoco anda haciendo de la siguiente manera, como dice en el [apunte](https://docs.google.com/document/d/1VYBey56M0UU6C0689hAClAvF9ILE6E7nKIuOqrRJnWQ/edit#heading=h.suvplv8njkpx), que fue por el cual yo me guie: `Enum<TipoPrenda> tipo = TipoPrenda.PANTALON;
    tipo.getCategoria();`

Si me dan una mano para saber que estoy haciendo mal buenisimo. Saludos!

**Comment by nanderson87** [comment, marked as answer]
Hola @Franpanozzo 

Dada una clase `Categoria`. Como la siguiente:

_Categoria.java_
```
public enum Categoria {
    PARTE_SUPERIOR;
}
```

Cualquiera de estas dos opciones te deberían funcionar.

_TipoPrenda.java - Opción A_
```
public enum TipoPrenda {
    REMERA {
        @Override
        public Categoria getCategoria() {
            return Categoria.PARTE_SUPERIOR;
        }
    };

    public abstract Categoria getCategoria();
}
```

_TipoPrenda.java - Opción B_
```
public enum TipoPrenda {
    REMERA(Categoria.PARTE_SUPERIOR),
    CAMISA(Categoria.PARTE_SUPERIOR);

    private Categoria categoria;

    TipoPrenda(Categoria categoria) {
        this.categoria = categoria;
    }

    public Categoria getCategoria() {
        return categoria;
    }
}
```

En lo personal me parece más simple la opción B.

**Reply by gastonprieto** [reply]
Por las dudas que no se llegue a apreciar en lo que te puso Nico, es que esta mal tipado tu método `(Enum<Categoria>)` debería ser: `Categoria`. Fijate que el ejemplo opción a) es que dice el apunte.

**Reply by Franpanozzo** [reply]
Perfecto, ahi lo corregi pero me estaria tirando el mismo error, _java: cannot find symbol
  symbol:   method getCategoria()
  location: variable tipoPrenda of type java.lang.Enum<domain.prendas.TipoPrenda>_

el codigo quedo asi:

```
public class Prenda {
  private Enum<TipoPrenda> tipoPrenda;

  public Enum<Categoria> getCategoria(){
    return tipoPrenda.getCategoria();
  }

}

public enum TipoPrenda {
  REMERA {
    //Cada tipo tiene una lista de que materiales se les permite
    @Override
    public Categoria getCategoria() {
      return Categoria.PARTESUP;
    }
  },
  PANTALON {
    @Override
    public Categoria getCategoria() {
      return Categoria.PARTEINF;
    }
  },
  ZAPATILLA {
    @Override
    public Categoria getCategoria() {
      return Categoria.CALZADO;
    }
  };

  public abstract Categoria getCategoria();
}
```

**Reply by gastonprieto** [reply]
@Franpanozzo Relee mi comentario, esta mal tipado el campo.

**Reply by Franpanozzo** [reply]
Perfecto ahi me anduvo, tenia entendido que cuando declaras un tipo de dato enum en una clase externa que los utiliza es Enum<nombreDelEnum> , eso lo hicieron nada mas para el pseudocodigo correcto?? o cuando lo haces??

**Reply by gastonprieto** [reply]
Es muy probable, te acordas donde lo viste?

**Reply by Franpanozzo** [reply]
Mmmm estoy casi seguro que en un pseudocodigo, pero es tanto material que no la verdad. Pero bueno ya quedo resolvida la duda que es lo importante, Muchas gracias!!!
